File: android/app/src/main/java/org/moodle/data/repository/MoodleAuthRepository.kt
package org.moodle.data.repository

import android.net.Uri
import com.google.gson.Gson
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.map
import org.moodle.core.model.ConnectionMode
import org.moodle.core.model.AuthState
import org.moodle.core.model.HtmlFeature
import org.moodle.core.model.MoodleError
import org.moodle.core.model.MoodlePublicConfig
import org.moodle.core.model.MoodleResult
import org.moodle.core.model.SiteAccount
import org.moodle.core.model.SiteCapabilities
import org.moodle.core.network.AjaxRequest
import org.moodle.core.html.HtmlMoodleDataSource
import org.moodle.core.html.HtmlMoodleException
import org.moodle.core.network.MoodleApi
import org.moodle.core.network.MoodleUrl
import org.moodle.core.security.PendingSso
import org.moodle.core.security.SecureCredentialStore
import org.moodle.core.security.SsoProtocol
import org.moodle.core.security.SsoSessionStore
import org.moodle.data.local.AppPreferences
import org.moodle.data.local.MoodleDao
import org.moodle.data.local.toDomain
import org.moodle.data.local.toEntity
import java.util.Locale
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

interface MoodleAuthRepository {
    val accounts: Flow<List<SiteAccount>>
    val activeAccount: Flow<SiteAccount?>
    suspend fun inspectSite(inputUrl: String): MoodleResult<MoodlePublicConfig>
    suspend fun login(config: MoodlePublicConfig, username: String, password: String): MoodleResult<SiteAccount>
    suspend fun reauthenticate(accountId: String, username: String, password: String): MoodleResult<SiteAccount>
    fun beginSso(config: MoodlePublicConfig): MoodleResult<String>
    suspend fun completeSso(callback: Uri): MoodleResult<SiteAccount>
    suspend fun activate(accountId: String)
    suspend fun remove(accountId: String)
}

@Singleton
class DefaultMoodleAuthRepository @Inject constructor(
    private val api: MoodleApi,
    private val dao: MoodleDao,
    private val tokenStore: SecureCredentialStore,
    private val ssoSessionStore: SsoSessionStore,
    private val appPreferences: AppPreferences,
    private val htmlDataSource: HtmlMoodleDataSource,
    private val gson: Gson,
) : MoodleAuthRepository {
    override val accounts: Flow<List<SiteAccount>> = dao.observeAccounts().map { rows ->
        rows.map { it.toDomain(gson) }
    }
    override val activeAccount: Flow<SiteAccount?> = combine(accounts, appPreferences.activeAccountId) { rows, activeId ->
        rows.firstOrNull { it.id == activeId } ?: rows.firstOrNull { it.isActive }
    }

    override suspend fun inspectSite(inputUrl: String): MoodleResult<MoodlePublicConfig> = safely {
        val baseUrl = MoodleUrl.normalize(inputUrl)
        val language = Locale.getDefault().language.take(8).ifBlank { "en" }
        val response = api.publicConfig(
            "$baseUrl/lib/ajax/service-nologin.php?info=tool_mobile_get_public_config&lang=$language",
            listOf(AjaxRequest(methodname = "tool_mobile_get_public_config")),
        ).firstOrNull() ?: error("The Moodle site returned an empty configuration")
        if (response.error) {
            val exception = response.exception
            throw MoodleRepositoryException(
                exception?.errorcode ?: "public_config_failed",
                exception?.message ?: "The Moodle public configuration could not be read",
            )
        }
        val config = response.data ?: error("The Moodle site returned an empty configuration")
        val canonical = MoodleUrl.normalize(config.httpswwwroot.ifBlank { config.wwwroot.ifBlank { baseUrl } })
        require(MoodleUrl.sameSite(baseUrl, canonical)) {
            "The Moodle canonical URL must stay on the entered site"
        }
        MoodlePublicConfig(
            canonicalUrl = canonical,
            siteName = config.sitename.ifBlank { canonical },
            mobileWebServiceEnabled = config.enablemobilewebservice == 1,
            loginType = config.typeoflogin,
            launchUrl = config.launchurl,
            showLoginForm = config.showloginform != 0,
        )
    }

    override suspend fun login(
        config: MoodlePublicConfig,
        username: String,
        password: String,
    ): MoodleResult<SiteAccount> = safely {
        if (config.connectionMode == ConnectionMode.NativeHtml) {
            return@safely createHtmlAccount(config, username, password)
        }
        val response = api.loginToken(
            "${config.canonicalUrl}/login/token.php?lang=${Locale.getDefault().language}",
            username.trim(),
            password,
        )
        val token = response.token ?: throw MoodleRepositoryException(
            response.errorcode ?: "login_failed",
            response.error ?: "Login failed",
        )
        createNativeAccount(config, token, response.privateToken, username.trim())
    }

    override suspend fun reauthenticate(
        accountId: String,
        username: String,
        password: String,
    ): MoodleResult<SiteAccount> = safely {
        val existing = dao.getAccount(accountId)?.toDomain(gson)
            ?: throw MoodleRepositoryException("account_missing", "The Moodle account no longer exists")
        require(existing.connectionMode == ConnectionMode.NativeHtml) { "Only HTML accounts use this sign-in flow" }
        val result = htmlDataSource.login(
            accountId,
            MoodlePublicConfig(existing.baseUrl, existing.siteName, false, 1, null, true),
            username.trim(),
            password,
        )
        val candidate = existing.copy(
            username = username.trim(),
            userId = result.identity.userId ?: existing.userId,
            fullName = result.identity.fullName ?: existing.fullName,
            siteName = result.identity.siteName,
            capabilities = SiteCapabilities(htmlFeatures = result.identity.features),
            authState = AuthState.Authenticated,
            moodleVersion = result.identity.moodleVersion,
            themeFamily = result.identity.themeFamily,
        )
        val refreshed = candidate.copy(capabilities = SiteCapabilities(htmlFeatures = validatedHtmlFeatures(candidate)))
        dao.upsertAccount(refreshed.toEntity(gson))
        activate(refreshed.id)
        refreshed
    }

    override fun beginSso(config: MoodlePublicConfig): MoodleResult<String> = runCatching {
        require(config.connectionMode == ConnectionMode.NativeApi) { "Mobile web services are disabled" }
        val passport = SsoProtocol.createPassport()
        ssoSessionStore.save(PendingSso(config.canonicalUrl, passport, nowEpochSeconds()))
        SsoProtocol.launchUrl(config.canonicalUrl, config.launchUrl, passport)
    }.fold(
        onSuccess = { MoodleResult.Success(it) },
        onFailure = { MoodleResult.Failure(it.asMoodleError()) },
    )

    override suspend fun completeSso(callback: Uri): MoodleResult<SiteAccount> = safely {
        val pending = ssoSessionStore.consume() ?: throw MoodleRepositoryException(
            "missing_sso_request",
            "No matching SSO request is pending",
        )
        val credentials = SsoProtocol.validateCallback(callback, pending, nowEpochSeconds())
        val inspected = inspectSite(credentials.baseUrl)
        val config = (inspected as? MoodleResult.Success)?.value ?: throw MoodleRepositoryException(
            "site_check_failed",
            (inspected as MoodleResult.Failure).error.message,
        )
        createNativeAccount(config, credentials.token, credentials.privateToken, null)
    }

    override suspend fun activate(accountId: String) {
        dao.setActiveAccount(accountId)
        appPreferences.setActiveAccountId(accountId)
    }

    override suspend fun remove(accountId: String) {
        dao.getAccount(accountId)?.toDomain(gson)?.takeIf { it.connectionMode == ConnectionMode.NativeHtml }
            ?.let { htmlDataSource.clearSession(accountId, it.baseUrl) }
        tokenStore.delete(accountId)
        appPreferences.clearActiveAccountIfMatches(accountId)
        dao.deleteAccount(accountId)
    }

    private suspend fun createNativeAccount(
        config: MoodlePublicConfig,
        token: String,
        privateToken: String?,
        enteredUsername: String?,
    ): SiteAccount {
        val siteInfo = callRest(
            config.canonicalUrl,
            token,
            "core_webservice_get_site_info",
            emptyMap(),
        ).asJsonObject
        val functions = siteInfo.getAsJsonArray("functions")
            ?.mapNotNull { it.asJsonObject.get("name")?.asString }
            ?.toSet()
            .orEmpty()
        val account = SiteAccount(
            id = UUID.randomUUID().toString(),
            baseUrl = config.canonicalUrl,
            siteName = siteInfo.string("sitename") ?: config.siteName,
            username = siteInfo.string("username") ?: enteredUsername,
            userId = siteInfo.long("userid"),
            fullName = siteInfo.string("fullname"),
            connectionMode = ConnectionMode.NativeApi,
            capabilities = SiteCapabilities(functions),
            isActive = true,
        )
        tokenStore.put(account.id, token, privateToken)
        dao.upsertAccount(account.toEntity(gson))
        activate(account.id)
        return account
    }

    private suspend fun createHtmlAccount(
        config: MoodlePublicConfig,
        username: String,
        password: String,
    ): SiteAccount {
        val accountId = UUID.randomUUID().toString()
        val result = try {
            htmlDataSource.login(accountId, config, username.trim(), password)
        } catch (error: Throwable) {
            htmlDataSource.clearSession(accountId, config.canonicalUrl)
            throw error
        }
        val candidate = SiteAccount(
            id = accountId,
            baseUrl = config.canonicalUrl,
            siteName = result.identity.siteName,
            username = username.trim(),
            userId = result.identity.userId,
            fullName = result.identity.fullName,
            connectionMode = ConnectionMode.NativeHtml,
            capabilities = SiteCapabilities(htmlFeatures = result.identity.features),
            authState = AuthState.Authenticated,
            moodleVersion = result.identity.moodleVersion,
            themeFamily = result.identity.themeFamily,
            isActive = true,
        )
        val account = candidate.copy(capabilities = SiteCapabilities(htmlFeatures = validatedHtmlFeatures(candidate)))
        dao.upsertAccount(account.toEntity(gson))
        activate(account.id)
        return account
    }

    private suspend fun validatedHtmlFeatures(account: SiteAccount): Set<HtmlFeature> {
        val features = account.capabilities.htmlFeatures
        if (account.userId == null) return features - MESSAGE_HTML_FEATURES
        val messagesAvailable = runCatching { htmlDataSource.conversations(account, 0, 1) }.isSuccess
        return if (messagesAvailable) features else features - MESSAGE_HTML_FEATURES
    }

    private suspend fun callRest(
        baseUrl: String,
        token: String,
        function: String,
        parameters: Map<String, String>,
    ) = api.restCall(
        "$baseUrl/webservice/rest/server.php",
        mapOf(
            "wstoken" to token,
            "wsfunction" to function,
            "moodlewsrestformat" to "json",
        ) + parameters,
    ).also { element ->
        if (element.isJsonObject && element.asJsonObject.has("exception")) {
            val error = element.asJsonObject
            throw MoodleRepositoryException(
                error.string("errorcode") ?: "webservice_error",
                error.string("message") ?: "Moodle web service error",
            )
        }
    }
}

private val MESSAGE_HTML_FEATURES = setOf(
    HtmlFeature.MessagesRead,
    HtmlFeature.MessagesSearch,
    HtmlFeature.MessagesSend,
    HtmlFeature.MessagesMarkRead,
)

class MoodleRepositoryException(val code: String, override val message: String) : Exception(message)

internal inline fun <T> safely(block: () -> T): MoodleResult<T> = runCatching(block).fold(
    onSuccess = { MoodleResult.Success(it) },
    onFailure = { MoodleResult.Failure(it.asMoodleError()) },
)

internal fun Throwable.asMoodleError(): MoodleError = when (this) {
    is MoodleRepositoryException -> MoodleError(code, message, code !in setOf("invalidtoken", "access_control_exception"))
    is HtmlMoodleException -> MoodleError(code, message, code !in setOf("invalid_credentials", "session_expired"))
    is IllegalArgumentException -> MoodleError("invalid_input", message ?: "Invalid input", false)
    else -> MoodleError("network_error", message ?: "Could not connect to Moodle")
}

internal fun com.google.gson.JsonObject.string(name: String): String? =
    get(name)?.takeUnless { it.isJsonNull }?.asString

internal fun com.google.gson.JsonObject.long(name: String): Long? =
    get(name)?.takeUnless { it.isJsonNull }?.runCatching { asLong }?.getOrNull()

private fun nowEpochSeconds(): Long = System.currentTimeMillis() / 1_000L

File: ios/ios/Data/MoodleAuthRepository.swift
import Foundation

@MainActor
protocol MoodleAuthRepository: AnyObject {
    func accounts() throws -> [SiteAccount]
    func activeAccount() throws -> SiteAccount?
    func inspectSite(_ inputURL: String) async throws -> MoodlePublicConfig
    func login(config: MoodlePublicConfig, username: String, password: String) async throws -> SiteAccount
    func reauthenticate(accountID: String, username: String, password: String) async throws -> SiteAccount
    func beginSSO(config: MoodlePublicConfig) throws -> URL
    func completeSSO(callback: URL) async throws -> SiteAccount
    func activate(accountID: String) throws
    func requireReauthentication(accountID: String) throws
    func remove(accountID: String) throws
}

@MainActor
final class DefaultMoodleAuthRepository: MoodleAuthRepository {
    private let transport: MoodleTransport
    private let store: MoodleStore
    private let keychain: KeychainStore
    private let html: HTMLMoodleDataSource
    private let mapper: MoodleMapper

    init(transport: MoodleTransport, store: MoodleStore, keychain: KeychainStore, html: HTMLMoodleDataSource, mapper: MoodleMapper) {
        self.transport = transport; self.store = store; self.keychain = keychain; self.html = html; self.mapper = mapper
    }

    func accounts() throws -> [SiteAccount] { try store.accounts() }
    func activeAccount() throws -> SiteAccount? { try store.activeAccount() }

    func inspectSite(_ inputURL: String) async throws -> MoodlePublicConfig {
        let baseURL = try MoodleURL.normalize(inputURL)
        return try await transport.publicConfig(baseURL: baseURL, language: Locale.current.language.languageCode?.identifier ?? "en")
    }

    func login(config: MoodlePublicConfig, username: String, password: String) async throws -> SiteAccount {
        let cleanedUsername = username.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !cleanedUsername.isEmpty, !password.isEmpty else { throw MoodleError(code: "invalid_credentials", message: String(localized: "error.login.required"), isRecoverable: false) }
        if config.connectionMode == .nativeHtml {
            guard config.showLoginForm else { throw MoodleError(code: "sso_html_unsupported", message: String(localized: "error.sso.html"), isRecoverable: false) }
            return try await createHTMLAccount(config: config, username: cleanedUsername, password: password)
        }
        let response = try await transport.token(baseURL: config.canonicalURL, username: cleanedUsername, password: password)
        guard let token = response.token else { throw MoodleError(code: response.errorCode ?? "login_failed", message: response.error ?? String(localized: "error.login.credentials"), isRecoverable: false) }
        return try await createAPIAccount(config: config, token: token, privateToken: response.privateToken, enteredUsername: cleanedUsername)
    }

    func reauthenticate(accountID: String, username: String, password: String) async throws -> SiteAccount {
        guard var existing = try store.account(id: accountID) else { throw MoodleError(code: "account_missing", message: String(localized: "error.account.missing"), isRecoverable: false) }
        if existing.connectionMode == .nativeHtml {
            let config = MoodlePublicConfig(canonicalURL: existing.baseURL, siteName: existing.siteName, mobileWebServiceEnabled: false, loginType: 1, launchURL: nil, showLoginForm: true)
            let result = try await html.login(accountID: existing.id, config: config, username: username.trimmed, password: password)
            existing.username = username.trimmed; existing.userID = result.identity.userID ?? existing.userID
            existing.fullName = result.identity.fullName ?? existing.fullName; existing.siteName = result.identity.siteName
            existing.capabilities = SiteCapabilities(htmlFeatures: result.identity.features)
            existing.authenticationState = .authenticated; existing.moodleVersion = result.identity.moodleVersion
            existing.themeFamily = result.identity.themeFamily
            try store.saveAccount(existing); try store.activateAccount(existing.id); return existing
        }
        let response = try await transport.token(baseURL: existing.baseURL, username: username.trimmed, password: password)
        guard let token = response.token else { throw MoodleError(code: response.errorCode ?? "login_failed", message: response.error ?? String(localized: "error.login.credentials"), isRecoverable: false) }
        let info = try mapper.siteInfo(try await transport.rest(baseURL: existing.baseURL, token: token, function: "core_webservice_get_site_info"), fallback: existing.siteName)
        existing.username = info.username ?? username.trimmed; existing.userID = info.userID; existing.fullName = info.fullName
        existing.siteName = info.siteName; existing.capabilities = SiteCapabilities(functions: info.functions)
        existing.authenticationState = .authenticated
        try keychain.saveCredentials(StoredCredentials(token: token, privateToken: response.privateToken), accountID: existing.id)
        try store.saveAccount(existing); try store.activateAccount(existing.id); return existing
    }

    func beginSSO(config: MoodlePublicConfig) throws -> URL {
        guard config.connectionMode == .nativeApi else { throw MoodleError(code: "sso_html_unsupported", message: String(localized: "error.sso.html"), isRecoverable: false) }
        let pending = PendingSSO(baseURL: config.canonicalURL, passport: SSOProtocol.createPassport(), createdAt: Date())
        try keychain.savePendingSSO(pending)
        return try SSOProtocol.launchURL(config: config, passport: pending.passport)
    }

    func completeSSO(callback: URL) async throws -> SiteAccount {
        guard let pending = try keychain.consumePendingSSO() else { throw MoodleError(code: "missing_sso", message: String(localized: "error.sso.missing"), isRecoverable: false) }
        let credentials = try SSOProtocol.validate(callback: callback, pending: pending)
        let config = try await inspectSite(credentials.baseURL.absoluteString)
        return try await createAPIAccount(config: config, token: credentials.token, privateToken: credentials.privateToken, enteredUsername: nil)
    }

    func activate(accountID: String) throws { try store.activateAccount(accountID) }

    func requireReauthentication(accountID: String) throws {
        guard var account = try store.account(id: accountID) else { return }
        account.authenticationState = .reauthenticationRequired
        try store.saveAccount(account)
        try store.activateAccount(accountID)
    }

    func remove(accountID: String) throws {
        if try store.account(id: accountID)?.connectionMode == .nativeHtml { html.clear(accountID: accountID) }
        try keychain.deleteCredentials(accountID: accountID)
        try store.deleteAccount(accountID)
        try FileManager.default.removeItem(at: FileCache.directory(accountID: accountID))
    }

    private func createAPIAccount(config: MoodlePublicConfig, token: String, privateToken: String?, enteredUsername: String?) async throws -> SiteAccount {
        let info = try mapper.siteInfo(try await transport.rest(baseURL: config.canonicalURL, token: token, function: "core_webservice_get_site_info"), fallback: config.siteName)
        let account = SiteAccount(id: UUID().uuidString, baseURL: config.canonicalURL, siteName: info.siteName,
                                  username: info.username ?? enteredUsername, userID: info.userID, fullName: info.fullName,
                                  connectionMode: .nativeApi, capabilities: SiteCapabilities(functions: info.functions), isActive: true)
        try keychain.saveCredentials(StoredCredentials(token: token, privateToken: privateToken), accountID: account.id)
        try store.saveAccount(account); try store.activateAccount(account.id); return account
    }

    private func createHTMLAccount(config: MoodlePublicConfig, username: String, password: String) async throws -> SiteAccount {
        let id = UUID().uuidString
        do {
            let result = try await html.login(accountID: id, config: config, username: username, password: password)
            var features = result.identity.features
            let candidate = SiteAccount(id: id, baseURL: config.canonicalURL, siteName: result.identity.siteName,
                                        username: username, userID: result.identity.userID, fullName: result.identity.fullName,
                                        connectionMode: .nativeHtml, capabilities: SiteCapabilities(htmlFeatures: features),
                                        moodleVersion: result.identity.moodleVersion, themeFamily: result.identity.themeFamily, isActive: true)
            let messagingAvailable: Bool
            if candidate.userID == nil {
                messagingAvailable = false
            } else {
                messagingAvailable = (try? await html.conversations(account: candidate, offset: 0, limit: 1)) != nil
            }
            if !messagingAvailable {
                features.subtract([.messagesRead, .messagesSearch, .messagesSend, .messagesMarkRead])
            }
            var account = candidate; account.capabilities = SiteCapabilities(htmlFeatures: features)
            try store.saveAccount(account); try store.activateAccount(account.id); return account
        } catch { html.clear(accountID: id); try? keychain.deleteCredentials(accountID: id); throw error }
    }
}

private extension String { var trimmed: String { trimmingCharacters(in: .whitespacesAndNewlines) } }
